stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/search/SearchViewModel.kt

```diff
@@ -5,57 +5,116 @@ import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
 import io.getstream.chat.android.client.ChatClient
 import io.getstream.chat.android.client.api.models.SearchMessagesRequest
+import io.getstream.chat.android.client.call.await
+import io.getstream.chat.android.client.logger.ChatLogger
 import io.getstream.chat.android.client.models.Filters
 import io.getstream.chat.android.client.models.Message
-import io.getstream.chat.android.client.models.User
-import io.getstream.chat.android.client.models.image
-import io.getstream.chat.android.client.models.name
-import java.util.Date
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
 
 public class SearchViewModel : ViewModel() {
 
-    private val _results: MutableLiveData<List<Message>> = MutableLiveData()
-    public val results: LiveData<List<Message>> = _results
+    public data class State(
+        val query: String,
+        val canLoadMore: Boolean,
+        val results: List<Message>,
+        val isLoading: Boolean,
+        val isError: Boolean,
+    )
+
+    private companion object {
+        const val QUERY_LIMIT = 30
+
+        val INITIAL_STATE = State(
+            query = "",
+            results = emptyList(),
+            isLoading = false,
+            isError = false,
+            canLoadMore = true,
+        )
+    }
+
+    private val scope = CoroutineScope(DispatcherProvider.Main + SupervisorJob())
+
+    override fun onCleared() {
+        super.onCleared()
+        scope.cancel()
+    }
+
+    private val _state: MutableLiveData<State> = MutableLiveData(INITIAL_STATE)
+    public val state: LiveData<State> = _state
+
+    private val logger = ChatLogger.get("SearchViewModel")
 
     public fun setQuery(query: String) {
-        fetchDummyResults(query)
+        scope.launch {
+            _state.value = State(
+                query = query,
+                results = emptyList(),
+                isLoading = true,
```

Comment:
Why do you prefer a combination of boolean flags instead of type hierarchy?

Reply:
The _loading_ state here is not distinct from the _loaded_ state. The View has existing data to display, as well as a query and paging information at the same time as the loading is happening.


Relevant: [my thoughts on designing single View states](https://zsmb.co/designing-and-working-with-single-view-states-on-android/), particularly [this section](https://zsmb.co/designing-and-working-with-single-view-states-on-android/#data-classes-for-mixed-states)